libraries/js/text/regex.effekt

```diff
@@ -1,33 +1,186 @@
-    case None() => None()
-    case Some(v) => Some(Match(v.matched, v.index))
-  }
 
-// internal representation { matched: String, index: Int }
+/**
+ * Constructs a Regex object from a pattern string and optional flags.
+ *
+ * @param pattern The regular expression pattern as a string.
+ * @param flags A string containing any combination of the following flags:
+ *   - `g`: Global search (find all matches rather than stopping after the first match)
+ *   - `i`: Case-insensitive search
+ *   - `m`: Multi-line search (^ and $ match start/end of each line)
+ *   - `d`: Generate indices for substring matches
+ *
+ * @return A Regex object that can be used with the `exec` function.
+ *
+ * Example usage:
+ * ```effekt
+ * val simpleRegex = regex("\\w+")
+ * val complexRegex = regex("(\\w+)\\s(\\d+)", "gi")
+ * ```
+ *
+ * Note: The `d` flag is required for extracting substring indices of groups.
+ * If not present, the indices of capture groups will always be `None`.
+ */
+def regex(pattern: String, flags: String): Regex = {
+    extRegex(pattern, flags)
+}
+
+extern pure def extRegex(str: String): Regex = js "new RegExp(${str})"
+extern pure def extRegex(str: String, flags: String): Regex = js "new RegExp(${str}, ${flags})"
+
+/**
+ * Executes the regex against the given string.
+ *
+ * @param reg The Regex object to execute.
+ * @param str The string to match against.
+ * @return Option[Match] A Some(Match) if a match is found, None otherwise.
+ *
+ * The returned Match object contains:
+ * - content: The full matched string.
+ * - start: The starting index of the match.
+ * - end: The ending index of the match.
+ * - groups: A list of Group objects for each capturing group.
+ *
+ * Each Group object contains:
+ * - content: The matched substring for this group.
+ * - index: An Option[Range] with start and end indices if the 'd' flag is set, None otherwise.
+ *
+ * Note: This function is stateful when the global flag 'g' is set on the Regex object.
+ * Subsequent calls with the same Regex and string will return the next match in the string.
+ *
+ * Example usage:
+ * ```effekt
+ * val rx = regex("(\\w+)\\s(\\d+)", "g")
+ * val text = "Hello 123 World 456"
+ * while (true) {
+ *   exec(rx, text) match = {
+ *      Some(m) =>
+ *          println("content: " ++ m.content ++ "range: " ++ m.start.show() ++ ", " ++ m.end.show()))
+ *      None() => break
+ * }
+ * ```
+ */
+def exec(reg: Regex, str: String): Option[Match] = {
+    def toGroup(g: GroupMatch): Group = Group(
+        content(g),
+        zip(undefinedToOption(start(g)), undefinedToOption(end(g))) { (s, e) => Range(s, e)}
+    )
+
+    undefinedToOption(reg.unsafeExec(str)) match {
+        case None() => None()
+        case Some(v) =>
+        val fullMatch = fullMatch(v)
+        Some(Match(
+            fullMatch.content,
+            fullMatch.start,
+            fullMatch.end,
+            v.subGroups.toList().map{toGroup}
+        ))
+    }
```

Comment:
Nitpick: since you were asking about style, I'd personally ANF the scrutinee out and then use `.map` on the result. I also like UFCS perhaps a bit too much. (suggestion untested)
```suggestion
    val regexMatch: Option[RegexMatch] = reg.unsafeExec(str).undefinedToOption
    regexMatch.map { v =>
      val groupMatch = v.fullMatch
      Match(
        groupMatch.content,
        groupMatch.start,
        groupMatch.end,
        v.subGroups.toList.map{toGroup}
      )
    }
```